File: 프로그래머스/2/87946. 피로도/피로도.py
import itertools
def solution(k, dungeons):
    answer = 0
    for ds in itertools.permutations(dungeons, len(dungeons)):
        a = 0
        nk = k
        for mf, sf in ds: 
            if nk >= mf:
                nk -= sf
                a += 1
        answer = max(answer, a)
    return answer

File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;
class Solution {
    int n, answer;
    boolean visited[];
    int[][] dungeon;
    
    void backtrack(int cnt, int k) {
        if(cnt == n || k < dungeon[cnt][0] ) {
            answer = Math.max(cnt, answer);
        }
        
        for(int i = 0; i<n; i++) {
            if(k >= dungeon[i][0]) {
               if(!visited[i]){
                    visited[i] = true;
                    backtrack(cnt+1, k-dungeon[i][1]);
                    visited[i] = false;
                } 
            }
        }
    }
    
    public int solution(int k, int[][] dungeons) {
        answer = -1;
        n = dungeons.length;
        visited = new boolean[n];
        dungeon = dungeons;
        backtrack(0, k);
        return answer;
    }
}